src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.Analyzers/TypeUpgradeAnalyzer.cs

```diff
@@ -0,0 +1,174 @@
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.TypeUpgradeDescription), Resources.ResourceManager, typeof(Resources));
+
+        private static readonly DiagnosticDescriptor Rule = new(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        /// <summary>
+        /// Initializes the analyzer by registering analysis callback methods.
+        /// </summary>
+        /// <param name="context">The context to use for initialization.</param>
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                // Load analyzer configuration defining the types that should be mapped.
+                _mappings = LoadMappings(context.Options.AdditionalFiles);
+
+                // If type maps are present, register syntax node actions to analyze for those types
+                if (_mappings.Any())
+                {
+                    // Register actions for handling both C# and VB identifiers
+                    context.RegisterSyntaxNodeAction(AnalyzeCSharpIdentifier, CS.SyntaxKind.IdentifierName);
+                    context.RegisterSyntaxNodeAction(AnalyzeVBIdentifier, VB.SyntaxKind.IdentifierName);
+                }
+            });
+        }
+
+        /// <summary>
+        /// Creates a type upgrade diagnsotic.
+        /// </summary>
+        /// <param name="location">The location the diagnostic occurs at.</param>
+        /// <param name="properties">Properties (including the name of the new identifier that code fix providers should substitute in) that should be included in the diagnostic.</param>
+        /// <param name="messageArgs">Arguments (the simple name of the identifier to be replaced and the full name of the identifier to replace it) to be used in diagnotic messages.</param>
+        /// <returns>A diagnostic to be shown to the user.</returns>
+        private static Diagnostic CreateDiagnostic(Location location, ImmutableDictionary<string, string?> properties, params object[] messageArgs)
+            => Diagnostic.Create(Rule, location, properties, messageArgs);
+
+        private void AnalyzeCSharpIdentifier(SyntaxNodeAnalysisContext context)
+        {
+            var identifier = (CSSyntax.IdentifierNameSyntax)context.Node;
```

Comment:
if we had an abstraction of IdentifierNameSyntax we could use a helper function to convert this to 1 standard type.

Obviously not a big help here but I think this is starting to look like a recurring problem. MemberAccessExpression was the other place I remember seeing this a lot and I think we actually have some helpers that use generics that are trying to solve the same problem.